### ISSUE TITLE
Project-level configuration files not found when run inside a Jupyer notebook

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#694]
In GitLab by @charlesfrye on Mar 7, 2019, 16:50

*Please describe how you installed Flake8*

I installed flake8 on Ubuntu 18.04 with pip inside a `virtualenv`

```
$ pip install flake8
```

*Please provide the exact, unmodified output of `flake8 --bug-report`*

```
{
  "dependencies": [
    {
      "dependency": "entrypoints",
      "version": "0.3"
    }
  ],
  "platform": {
    "python_implementation": "CPython",
    "python_version": "3.6.7",
    "system": "Linux"
  },
  "plugins": [
    {
      "is_local": false,
      "plugin": "mccabe",
      "version": "0.6.1"
    },
    {
      "is_local": false,
      "plugin": "pycodestyle",
      "version": "2.5.0"
    },
    {
      "is_local": false,
      "plugin": "pyflakes",
      "version": "2.1.1"
    }
  ],
  "version": "3.7.7"
}
```

*Please describe the problem or feature*

While using [a downstream repository](https://github.com/mattijn/pycodestyle_magic) to perform linting on the cells of Jupyter notebooks, I noticed that my configuration files were not being respected, even though they were respected when invoking `flake8` from the command line.

*If this is a bug report, please explain with examples (and example code) what you expected to happen and what actually happened.*

If I execute the following 

```python
from flake8.api import legacy
style_guide = legacy.get_style_guide()
style_guide.options
```

inside a Python terminal, the `ignore` attribute contains the `ignore`s from my project-level specification, as desired.

If I run the same lines in a Jupyter notebook in the same directory, only the default `ignore`s are included. `ignore`s specified in a `.flake8` file in that directory do not appear.

I did a little bit of digging to try and determine where the problem is occurring, and it seems to be somewhere between the `ConfigFileFinder` and the `Application`.
The code below is collected in a GitHub gist [here](https://gist.github.com/charlesfrye/d4876151a11c59fa35c0dc821f7c1171).

If I execute the internals of `get_style_guide` up to [37 in api.legacy](https://gitlab.com/pycqa/flake8/blob/master/src/flake8/api/legacy.py#L37)
in order to gain direct access to the `application`, like so:
```python
import flake8
from flake8.formatting import base as formatter
from flake8.main import application as app

application = app.Application()

application.parse_preliminary_options_and_args([])
flake8.configure_logging(
        application.prelim_opts.verbose, application.prelim_opts.output_file
    )
application.make_config_finder()
application.find_plugins()
application.register_plugin_options()
application.parse_configuration_and_cli([])
```

then, in a Jupyter notebook, `application.config_finder.local_config_files` returns an empty list, indicating that the local config files were not found, but the same procedure returns a non-empty list in a Python terminal.

But if, inside the Jupyter notebook, I then run the following:
```python
import flake8.options.config as config
cff = config.ConfigFileFinder(application.program_name, [], [])
cff.local_config_files()
```

the local config file *is* present. I was unable to dig any further to find out where the divergence in behavior is occurring.

Many thanks for providing an awesome tool! Your work is very much appreciated.

**Comment by asottile** [comment]
In GitLab by @charlesfrye on Mar 7, 2019, 16:52

changed the description

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 7, 2019, 16:55

Can you print os.getcwd() in both cases? Might contain a hint :thinking:

**Comment by asottile** [comment]
In GitLab by @charlesfrye on Mar 7, 2019, 16:59

`os.getcwd()` returns the same thing in both cases, the expected current directory.

I looked at the internals and it seems y'all tend to instead use `os.path.abspath` on `os.curdir`, which also returns the same in both cases.

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 7, 2019, 17:07

Interesting, I've never used jupyter myself before -- what's the minimum I need to set up to reproduce this?

**Comment by asottile** [comment]
In GitLab by @charlesfrye on Mar 7, 2019, 17:26

I was just able to recreate in a clean Python 3.6 virtualenv with only

```pip install jupyter flake8
jupyter notebook
```

and then, inside a fresh notebook, executing the code from the report/gist.

You'll need to run the command `jupyter notebook` in a terminal that can launch a web browser (default terminal on a GUI-enabled machine should be fine) in order to view the notebook.

I will toss up a GitHub repo with a ready-to-go notebook in it to save you some effort.

**Comment by asottile** [comment]
In GitLab by @charlesfrye on Mar 7, 2019, 17:31

Notebook [here](https://github.com/charlesfrye/flake8_518) will recreate from scratch in a suitable environment.

With maybe ten minutes more effort, I could get this up on a cloud instance and verify the bug there, letting you avoid the install process. However in that case the interface (including to terminals) would be strictly through Jupyter, so maybe a bad fit for your workflow.

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 7, 2019, 18:06

Tracing it a bit, it appears to be here where things go awry:

https://gitlab.com/pycqa/flake8/blob/88caf5ac484f5c09aedc02167c59c66ff0af0068/src/flake8/main/application.py#L125

Up a frame, we're at:

https://gitlab.com/pycqa/flake8/blob/88caf5ac484f5c09aedc02167c59c66ff0af0068/src/flake8/api/legacy.py#L30

This causes the preliminary args to be parsed as:
```
In [11]: application.prelim_args
Out[11]: ['/run/user/1000/jupyter/kernel-0b1b95f9-f268-4cca-b8ed-1d4b7d6c8b49.json']
```

This causes the config finder to look in that directory for configurations (?)

```
In [14]: application.config_finder.parent, application.config_finder.tail

Out[14]: ('/run/user/1000/jupyter/kernel-0b1b95f9-f268-4cca-b8ed-1d4b7d6c8b49.json',
          '/run/user/1000/jupyter/kernel-0b1b95f9-f268-4cca-b8ed-1d4b7d6c8b49.json')
```

So then we get to here:

https://gitlab.com/pycqa/flake8/blob/88caf5ac484f5c09aedc02167c59c66ff0af0068/src/flake8/options/config.py#L94-108

### hacky workaround

set `sys.argv[:] = ['flake8']` before making a style guide

### probably a fix

I *think* that line should probably say `argv if argv is not None else sys.argv[1:]` (or `sys.argv`? not sure exactly)

**Comment by asottile** [comment]
In GitLab by @charlesfrye on Mar 7, 2019, 18:25

I don't think that the **probably a fix** change would fix it -- `argv if argv is not None else sys.argv` evaluates to the same as `argv or sys.argv`.

It seems like the fundamental problem is that Jupyter has its own `sys.argv` dating back to when the kernel was launched from the command line. This breaks the assumption of `flake8` that `sys.argv` is intended for its consumption.

Would it be reasonable to wrap in a `if __name__ == '__main__'` before pulling in `sys.argv`? Then in the `else`, you'd need to populate with some reasonable default, as in the `sys.argv[:] = ['flake8']` hack. As an outsider, not clear whether that's sensible or not.

Alternatively, this could be fixed downstream by the folks using `flake8` in a Jupyter context by them always providing an explicit `argv`. I can communicate that to them as an issue if that's the preferred route.

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 7, 2019, 18:29

It's different in this case because an empty list is explicitly passed in, so it would use the empty list of arguments

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 7, 2019, 18:34

in fact, here it is working as intended :)

![](https://i.fluffy.cc/r7LfnksCP8G0JbDKzCZc6RxzqDmMT4R1.png)

It's unclear if that has negative side-effects, if you want I can make the PR or you can (if you want to contribute :D) -- either way I can help write a test to demonstrate this issue

**Comment by asottile** [comment]
In GitLab by @charlesfrye on Mar 7, 2019, 18:43

I see, so this is an unintended behavior of `parse_preliminary_options_and_args` when given an empty list that is only detectable in the case where `sys.argv` is something unexpected.

I'd love the chance to contribute! Is it possible to PR to GitHub instead? I'm just more familiar with the API there. If not, I can figure GL out.

Help with the test-writing would be much appreciated!

**Comment by asottile** [comment]
In GitLab by @charlesfrye on Mar 7, 2019, 19:08

First thought regarding the test: could it be as narrow as

```python
app.parse_preliminary_options_and_args([]) == app.option_manager.parse_known_args([])[1]
```

or should it reproduce the whole chain of pulling from `sys.argv` when that's inappropriate?

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 7, 2019, 19:26

It would be preferable to PR here, but if you make a PR to github and tag me (`@asottile` there as well!) I can manually merge once we get it to a good place.  Definitely understand the familiarity of tools issues, I spend most of my time on github too :)

The nice thing is with the suggested fix the current tests pass, the only change would be to write a test that would exercise this behaviour.

An easy one that would demonstrate the behaviour you want (more of an integration-level test) would be to do something like:

```python
with mock.patch.object(sys, 'argv', ['myexe', '/path/to/foo']):
    # initialize an application
    # call the method which initializes the preliminary options with []
    # make sure that it doesn't use sys.arv
```

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 7, 2019, 19:37

(I wrote that up before seeing your response, I can respond in the morning, sorry)

**Comment by asottile** [comment]
In GitLab by @charlesfrye on Mar 7, 2019, 19:43

No worries! Your suggestion is more explicit.

Here's what I've got for a test, at the bottom of`tests/unit/test_application.py`

```python
def test_prelim_opts_handles_empty(application):
    """Verify empty argv lists are handled correctly."""
    irrelevant_args = ['myexe', '/path/to/foo']
    with mock.patch.object(sys, 'argv', ):
        application.parse_preliminary_options_and_args([])

        assert application.prelim_args != irrelevant_args
```

It seems the other tests reuse the `application` from the `@pytests.fixture` rather than creating one in each test.

Also, for the change to [L125](https://gitlab.com/pycqa/flake8/blob/88caf5ac484f5c09aedc02167c59c66ff0af0068/src/flake8/main/application.py#L125), did you do `[1:]` or `[:]` at the end of the line?

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 7, 2019, 19:47

I went with this:

```
        args = (argv if argv is not None else sys.argv)[:]
```

I assume that `parse_known_args` has some magic to make `sys.argv` handled correctly and/or it is dead code

I think your test is missing `, irrelevant_args` in the patch but otherwise looks good :thumbsup:  (I'm sad that gitlab tab completes `:thumbs` to `:thumbsdown:` so take solace in knowing I typed out `:thumbsup:` manually)

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Mar 8, 2019, 05:22

> I assume that `parse_known_args` has some magic to make `sys.argv` handled correctly and/or it is dead code

`parse_known_args` is absolutely not dead code unless things have changed drastically since 3.7.0. We have to double-parse command-line args for several reasons. The primary reason being that there are flags that affect things like logging and plugin loading. So we create a parser with the "core" flags (the ones we define) then we add any new flags we find as a result of our updated understanding of the world.

If we were on Argparse instead of Optparse then we could just use Argparse's `parse_known_args` to do this instead of implementing it ourselves. That said, our `OptionManager` has its own API so we could now swap out `argparse` for `optparse` without plugin developers being any the wiser.

@asottile what parts of the implementation do you need additional documentation about? I feel like we've discussed "potentially dead code" a few times and I'd like to ensure you and any other maintainer we add has the resources necessary to be confident in what is happening inside Flake8.

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 8, 2019, 07:48

I wasn't sure about the use of `sys.argv` ever being possible here. My quick glance last night showed callsites passing in a variable there and potentially none which pass `None` or no argument (which would make me believe that the second half of that `else` isn't hit and therefore dead).

I wasn't confident I could determine whether that was used or not as the test suite didn't respond to me changing it, I plan to look more into it when I have the time

My gut feeling for it being dead is the pattern I usually see is `argv or sys.argv[1:]` and if it isn't dead I suspect argparse has special code to handle both `sys.argv` and `sys.argv[1:]` that I've never experienced before despite spending lots of time with argparse

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 8, 2019, 09:09

Ok, I traced it through, this is the case where `argv=None`: https://gitlab.com/pycqa/flake8/blob/88caf5ac484f5c09aedc02167c59c66ff0af0068/src/flake8/__main__.py#L4

There's also code in that function that handles stripping the program name: https://gitlab.com/pycqa/flake8/blob/88caf5ac484f5c09aedc02167c59c66ff0af0068/src/flake8/main/application.py#L140-141

Though I think that means that it will strip the first argument in any case where `sys.argv` is not used -- looks like this applies to a few callsites:

- https://gitlab.com/pycqa/flake8/blob/88caf5ac484f5c09aedc02167c59c66ff0af0068/src/flake8/main/git.py#L46
- https://gitlab.com/pycqa/flake8/blob/88caf5ac484f5c09aedc02167c59c66ff0af0068/src/flake8/main/mercurial.py#L39

**Comment by asottile** [comment]
In GitLab by @charlesfrye on Mar 11, 2019, 14:08

mentioned in commit 16ca17388ab4a8136617390329328e1a13219f84

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 11, 2019, 14:09

mentioned in merge request !310

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 11, 2019, 15:00

closed via merge request !310

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 11, 2019, 15:00

closed via commit fadedefae2df9fa6dff624fab733d717aea5ca0e

**Comment by asottile** [comment]
In GitLab by @asottile on Mar 11, 2019, 15:00

@charlesfrye Thanks again for the patch and the report :pray:
